sonarts-core/src/symbols/collectionUtils.ts

```diff
@@ -0,0 +1,217 @@
+import { SymbolTable, Usage, UsageFlag } from "./table";
+import * as ts from "typescript";
+import { firstAncestor, COMPOUND_ASSIGNMENTS } from "../utils/navigation";
+import {
+  is,
+  isBinaryExpression,
+  isElementAccessExpression,
+  isCallExpression,
+  isPropertyAccessExpression,
+  isNewExpression,
+  isArrayLiteralExpression,
+  isAssignment,
+} from "../utils/nodes";
+
+const writingMethods = new Set([
+  "copyWithin",
+  "fill",
+  "pop",
+  "push",
+  "reverse",
+  "shift",
+  "sort",
+  "splice",
+  "unshift",
+  "clear",
+  "delete",
+  "set",
+  "add",
+]);
+
+const WRITE_ARRAY_PATTERNS: ((statement: ts.ExpressionStatement, usage: Usage) => boolean)[] = [
+  isElementWrite,
+  isVariableWrite,
+  isWritingMethodCall,
+];
+
+/**
+ * Returns an array of pair symbol-declaration storing collections.
+ * Parameters, exported/imported and type related symbols are filtered out.
+ */
+export function getCollectionSymbols(symbols: SymbolTable, program: ts.Program) {
```

Comment:
all logic here is so specific (so many things are filtered out). Can we move this class to `rules` dir?